### ISSUE TITLE
[Proposal] Stop requiring /admin/ folder rename

### ISSUE THREAD

**Issue by drbyte** [zencart/zc-v1-series#404]
Feedback welcome ...

Proposal: Instead of throwing a roadblock up if the /admin/ folder is not renamed, instead just display a warning message inside the admin after logging in.

**Comment by wiztechinc** [comment]
A roadblock it may be but it's dang effective. A message? Hmm, I got
clients who would start yelling at me and I've seen clients who almost
never log into admin - they fill orders from the order email!
ᐧ

Delia Wilson Lunsford
WizTech, Inc., (formerly Delia Wilson Design, LLC.)
434-202-4307
Terms and Conditions for working with WizTech, Inc.
(http://www.teamwiztech.com/terms-conditions.php)

**Comment by barco57** [comment]
Why would your clients yell at you Delia...shouldn't you have already renamed their admin folders for them...

**Comment by wiztechinc** [comment]
Oh, no, not because I did anything wrong!!

**Comment by zcwilt** [comment]
Just some background here, and reasons to support the issue.

A long time ago, ZC admin was pretty insecure :(
It had little in the way of xss/xsrf protection and was subject to some other horrible vulnerabilities.
Obfuscating the admin directory was, at the time a quick way of mitigating against this.

Since then the admin code has been security hardened (and tested thru multiple PA-DSS validations) 

The intent of this issue is not to stop store owners changing the admin directory, but to not `roadblock` the admin if they haven't.

There are a number of reasons why.

1. Security by obscurity is bad, and can lead to complacency in managing security issues.
2. Easier to document.
3. Moving to use shared configs/controllers for admin/store

**Comment by wiztechinc** [comment]
Easier to upgrade, I must admit, if the folder is named admin. One thing that not having a standard admin folder name is that it cuts down on folks trying to hack in. So it's easier on the server (I suspect) since they are just getting a 404 in response now. Don't know if that really makes a difference or not but since I host sites as well, I appreciate them not finding what they are looking for.

Since plugins and custom code can lessen security, do we really want to open up the possibility of hacking again? I assume zen cart admin can be made vulnerable by those who don't know any better.

**Comment by drbyte** [comment]
> @wiztechinc wrote: 
> open up the possibility of hacking again

Sigh. That choice of wording can be sorely misinterpreted by alarmists.

One thing you probably aren't aware of is that GoDaddy and other 1-click-install hosts have chosen their own "admin" folder name that's used on every install. So, while it's no longer called just "admin" (and therefore doesn't flag up the roadblock alert-page), it's really simple to guess a few common choices used by those hosts, and bam you're on the login page. So, since those hosts have totally broken the (albeit minor) protection offered by security-by-obscurity, and have been doing so for several years, there's little reason to keep "forcing" the rename on newcomers. 

Of course, since doing a folder rename is a simple task and no longer requires touching config files, one can always rename their admin folder anytime they want, and thus benefit from the insulation that offers.

**Comment by wiztechinc** [comment]
Too true. That's always bothered me with the automated installs. I see maybe a third or less of sites with zc_admin.  A third come with their own name and the rest are my own creation. Hard to know how widespread that is.

**Comment by zcwilt** [comment]
> Since plugins and custom code can lessen security, do we really want to open up the possibility of hacking again? I assume zen cart admin can be made vulnerable by those who don't know any better.

Sure, of course it's possible to write contributions that totally compromise security. However that's true  whether the admin is renamed or not, and as I mentioned before relying on the obscurity leads to complacency.

We are moving towards having a much better API in admin code, such that writing contributions relies more on core code than custom code (e.g. https://github.com/zencart/zc-v1-series/issues/63 etc.) 

Of course we can't force people to user these new API's, but we hope the most popular contributions will update for 1.6.

**Comment by wiztechinc** [comment]
Sounds like work for me!! But anything that betters zen cart, I'm all for.

**Comment by DivaVocals** [comment]
I agree with @drbyte warning versus roadblock..

**Comment by lhungil** [comment]
:+1: on warning instead of roadblock. Especially if the new Zen Cart installer requires one to log into the Zen Cart admin to "finish" the installation.

**Comment by scottcwilson** [comment]
Am I the only one who will vote for the status quo?  If it's not admin, and the attacker isn't serious, they'll just move on - this is at least something. Why discard it for no reason?

**Comment by wiztechinc** [comment]
I agree, Scott. I understand the reasoning behind this but if I have to choose I go for roadblock.

**Comment by barco57** [comment]
I am 60/40 on change it versus status quo.....would the warning in the admin be perisitent? would it be shown once and then go away for a period of time? Would it be better to have a popup modal in the admin the first time someone logs into a new install with "Welcome to your new Zen Cart" and a list of probable things they should do first, link to forum, best practices, picture of Chris...lol, ok, maybe not a picture of Chris, just some sort of important stuff with a "don't show again" check box or something....

**Comment by drbyte** [comment]
What, you don't like my mug shot? :crying_cat_face:

**Comment by syntaxerror** [comment]
I've seen some system having a config setting for adding a "secret keyword", so you get a admin URL like "mydomain.tld/admin?mykeyword", and if it's missing there is a redirect to frontend happening.

Could it be something to consider instead?

**Comment by jtwebsitedesign** [comment]
I may be a little late to this conversation, just wanted to chime in. I agree that it's a good idea to keep the added layer of security, and I think strictly enforcing the rename would avoid any oversight.

What about simplifying the admin rename process? Perhaps adding a field on the main setup page called "Admin URL" so the user can set it there, and it auto-renames the directory as part of the install process. Then there's zero possibility that someone would setup a cart without renaming, and no annoying roadblock/message either.

I like syntaxerror's idea too about the secret keyword. That would make it easier to maintain files since they're always in the "admin" directory, but you'd also have the URL login security. Win-win.

Perhaps a combination of both of these ideas would be way more user friendly, and also be more secure?

**Comment by zcwilt** [comment]
Going to close this as the consensus seems to be stick with what we have.

**Comment by dakanji** [comment]
Sorry to dig this up but it would really be great to go with the proposal by @drbyte and @zcwilt  as it just makes sense.

Currently, with doing quite a few test installations and other development things, the renaming is just a PITA with next to zero real benefit.

**Comment by syntaxerror** [comment]
@zcwilt I do actually wonder a bit about your counting on people supporting vs being against, or are you and @drbyte against the change - is there more than three people being against here? ;)

**Comment by scottcwilson** [comment]
I would vote to reject that change.  Changing from "/admin" helps discourage script kiddies from stupid attacks that waste bandwidth.  For development, keep /admin and use the `ADMIN_BLOCK_WARNING_OVERRIDE` flag in configure.php.

**Comment by syntaxerror** [comment]
You can use htaccess, firewall etc for that too ... anyhow you already "voted" above, and is included in my count - so I still wonders about this counting thing. :)

**Comment by dakanji** [comment]
Surely saving bandwidth can't be a serious consideration here ... or is it? Really??

**Comment by syntaxerror** [comment]
Bandwidth is influenced if there is some "attaches" going on, but normally you must have a really bad hosting if that makes issues for you, when "script kiddies" plays around.

**Comment by scottcwilson** [comment]
For anyone who finds the rename troubling, in 160, you can edit zc_install/includes/localConfig.php and add 

`define('DEVELOPER_MODE', true); `

and it will no longer rename.

**Comment by drbyte** [comment]
Update:
The better way to skip admin folder renaming when working in a local dev environment, which also bypasses git add hassles, is:

create a new file called: `/zc_install/includes/extra_configures/dev-developer-mode.php`

containing:
```php
<?php
define('DEVELOPER_MODE', true); 
```

**NOTE: This should NOT be present for production site installs.**

**Comment by dakanji** [comment]
... or install the Habitat/Vagrant environment which the dev team has allowed for in the installation code. 

I have personally decided to suck it and deal with it as the dev thing is not the only/key attraction for proposing the change but rather agreeing with well elucidated reasons outlined by @zcwilt & @drbyte.

I used to manage the Gallery2/Zencart bridge. Not sure how something like that will work these days when the catalog side can't even share stuff with the admin side.

Seems to me as a good example of when devs have a very good idea but it never sees the light of day after putting it to we, the masses. 

Anyway, suppose you have to take the smooth with the rough so considering this unfortunately firmly closed and moving on.

**Comment by drbyte** [comment]
> when the catalog side can't even share stuff with the admin side

Not sure what you mean by this. There's more sharing happening now between catalog/admin than ever before.

**Comment by syntaxerror** [comment]
Might be, but wasn't it a reason to support the change, that it could be even better?!
Ref https://github.com/zencart/zencart/issues/404#issuecomment-98811613
`3. Moving to use shared configs/controllers for admin/store`

**Comment by drbyte** [comment]
I gather rune and dayo are making these comments in an attempt to lobby for this issue to be re-opened? Is that correct? Because otherwise your comments make no sense.

**Comment by scottcwilson** [comment]
To me, I think we are in good shape - there is a very reasonable way of getting around the rename for developer-only local sites (your update to my post + fix #1123).

**Comment by dakanji** [comment]
@drbyte I did make my original comment https://github.com/zencart/zencart/issues/404#issuecomment-232551853 for the issue to be reopened/reconsidered (thought it would might be reopened automatically) based on the developer mode thing and the list of strong reasons you and @zcwilt set out originally.

The fix by @scottcwilson does mitigate the developer mode issue :thumbsup: but coming back to a full reconsideration, I thought a summation of things was:

- The rename process is in place based on a perception that it improves security in a significant manner and also apparently saves a precious resource in bandwidth.
- The Dev Team's view, as set out earlier in this thread, is that the security perception is not quite the same as reality
- The bandwidth angle is better left alone.
- Removing the rename thing will allow certain beneficial procedures to be put in place

Given that, it seemed to me that the process should be removed. 

However, I can see the potential for a knee jerk, gut reaction from users of this "leaving them open to hacking" or "allowing script kiddies to chew up their bandwidth" or "make them vulnerable to DDOS attacks" and can understand why the Dev Team might not want to deal with this. So, I have accepted that the rename thing is here to stay for good and no longer pushing/hoping for reconsideration.  

The issue can stay closed from my side.

**Comment by syntaxerror** [comment]
@scottcwilson It's not about devs only as the history in this clearly shows, but hey, I give up on the narrowed thinking and move on. Have only 5 regular clients left on ZC, it's stable, but the old thinking and fight against changes is killing it.

**Comment by drbyte** [comment]
Wow ... Clearly this "rename the admin folder" is a seriously important issue to you!

I never imagined that the delay in doing it would be considered "old thinking and fight against changes". 

> @syntaxerror wrote "there must have been quite much lobbing going on behind the scene"

Haha! I wish!
There's no "lobbying" going on.
@zcwilt and @ajeh and I have a voice conversation several times a week, and after we closed this issue we decided we will eventually do it anyway, but probably not until after some improvements are made to the admin message alerting. There are bigger more urgent priorities at the moment, hence we closed the ticket.

Evidently that was confusing to you all .... but more than that it sure has stirred up quite the side discussion of accusations. 

Come on guys. You're bigger than that. If you really think things "should evolve" in certain areas, open issues for those things, rather than just making generalized negative comments in unrelated areas.

**Comment by syntaxerror** [comment]
Just replied to your somehow weird comment about lobbying ... This "issue"isolated is not very important, true, but the total is. Clearly some has more to say, and other are simply being ignored here (it's actually considered rude in most of the world). It's just the way it is.

> @zcwilt  wrote "Going to close this as the consensus seems to be stick with what we have."
 :-1: 1  

The comment posted on the close doesn't imply anything about this being postponed or anything, it's rather opposite. But nevermine, sorry hitting that nerve.
